File: src/org/genyris/interp/Runner.java
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.interp;

import java.io.IOException;
import java.io.Writer;

import org.genyris.core.Constants;
import org.genyris.core.Exp;
import org.genyris.format.IndentedFormatter;
import org.genyris.io.ConvertEofInStream;
import org.genyris.io.InStream;
import org.genyris.io.IndentStream;
import org.genyris.io.Parser;
import org.genyris.io.StringInStream;
import org.genyris.io.UngettableInStream;

public class Runner {

    public static Exp executeScript(Interpreter interp, StringInStream stream, Writer output) throws GenyrisException {
        InStream input = new UngettableInStream(new ConvertEofInStream(new IndentStream(new UngettableInStream(stream),
                false)));
        Parser parser = interp.newParser(input);
        IndentedFormatter formatter = new IndentedFormatter(output, 3, interp);
        Exp expression = null;
        Exp result = null;
        do {
            expression = parser.read();
            if (expression.equals(interp.getSymbolTable().internString(Constants.EOF)) ) {
                break;
            }
            result = interp.evalInGlobalEnvironment(expression);
            result.acceptVisitor(formatter);
            try {
                output.flush();
            } catch (IOException ignore) {}
        } while (true);
        return result;
    }
}

File: src/org/genyris/interp/Runner.kt
// Copyright 2008 Peter William Birch <birchb@genyis.org>
//
// This software may be used and distributed according to the terms
// of the Genyris License, in the file "LICENSE", incorporated herein by reference.
//
package org.genyris.interp

import org.genyris.core.Exp
import org.genyris.exception.GenyrisException
import org.genyris.format.Formatter
import org.genyris.format.IndentedFormatter
import org.genyris.io.*
import java.io.IOException
import java.io.Writer

object Runner {
    @kotlin.Throws(GenyrisException::class)
    fun executeScript(interp: Interpreter, stream: StringInStream?, output: Writer): Exp? {
        val input: InStream = UngettableInStream(
            ConvertEofInStream(
                IndentStream(
                    UngettableInStream(stream),
                    false
                )
            )
        )
        val parser = interp.newParser(input)
        val formatter: Formatter = IndentedFormatter(output, 3)
        var expression: Exp? = null
        var result: Exp? = null
        do {
            expression = parser.read()
            if (expression == interp.getSymbolTable().EOF()) {
                break
            }
            result = interp.evalInGlobalEnvironment(expression)
            result.acceptVisitor(formatter)
            try {
                output.flush()
            } catch (ignore: IOException) {
            }
        } while (true)
        return result
    }
}
